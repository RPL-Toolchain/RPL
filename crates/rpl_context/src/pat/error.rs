use std::sync::LazyLock;

use rpl_meta::symbol_table::{MetaVariableType, NonLocalMetaSymTab};
use rpl_parser::generics::Choice2;
use rpl_parser::pairs;
use rustc_errors::LintDiagnostic;
use rustc_lint::{Level, Lint};
use rustc_middle::mir::Body;
use rustc_span::{Span, Symbol};
use sync_arena::declare_arena;

use crate::pat::{ConstVarIdx, TyVarIdx};

use super::{LabelMap, Matched};

/// A dynamic error that can be used to report user-defined errors
///
/// See:
/// - [`rustc_errors::LintDiagnostic`].
/// - [`rustc_macros::diagnostics::lint_diagnostic_derive`].
/// - [`rustc_macros::LintDiagnostic`].
pub struct DynamicError {
    /// Primary message and its span.
    ///
    /// See [`rustc_errors::Diag::primary_message`].
    /// The primary message is the main error message that will be displayed to the user.
    primary: (String, Span),
    /// Label description, and the span of the label.
    ///
    /// See [`rustc_errors::Diag::span_label`].
    /// Labels are used to highlight specific parts of the code that are relevant to the error.
    labels: Vec<(String, Span)>,
    notes: Vec<(String, Option<Span>)>,
    helps: Vec<(String, Option<Span>)>,
    lint: &'static Lint,
}

impl LintDiagnostic<'_, ()> for DynamicError {
    fn decorate_lint(self, diag: &mut rustc_errors::Diag<'_, ()>) {
        let primary_message = self.primary.0;
        diag.primary_message(primary_message);
        for (label, span) in self.labels {
            diag.span_label(span, label);
        }
        for (help, span_help) in self.helps {
            if let Some(span_help) = span_help {
                diag.span_help(span_help, help);
            } else {
                diag.help(help);
            }
        }
        for (note, span_note) in self.notes {
            if let Some(span_note) = span_note {
                diag.span_note(span_note, note);
            } else {
                diag.note(note);
            }
        }
    }
}

impl DynamicError {
    pub const fn primary_span(&self) -> Span {
        self.primary.1
    }
    /// Also see [`rustc_session::declare_tool_lint!`].
    pub const fn lint(&self) -> &'static Lint {
        self.lint
    }
}

enum SubMsg<'i> {
    Str(&'i str),
    Ty(TyVarIdx),
    Const(ConstVarIdx),
}

impl<'i> SubMsg<'i> {
    fn parse(s: &pairs::diagMessageInner<'i, 0>, meta_vars: &NonLocalMetaSymTab) -> Vec<Self> {
        let mut msgs = Vec::new();
        for seg in s.iter_matched() {
            match seg {
                Choice2::_0(arg) => {
                    let (var_type, idx) = meta_vars
                        .get_type_and_idx_from_symbol(Symbol::intern(arg.MetaVariable().span.as_str()))
                        .unwrap();
                    match var_type {
                        MetaVariableType::Type => msgs.push(SubMsg::Ty(idx.into())),
                        MetaVariableType::Const => msgs.push(SubMsg::Const(idx.into())),
                        MetaVariableType::Place => panic!(
                            "Unexpected place meta variable in diagnostic message: {}",
                            arg.span.as_str()
                        ),
                    }
                },
                Choice2::_1(text) => {
                    msgs.push(SubMsg::Str(text.span.as_str()));
                },
            }
        }
        msgs
    }
}

pub(crate) struct DynamicErrorBuilder<'i> {
    /// Primary message and its span.
    primary: (Vec<SubMsg<'i>>, &'i str),
    /// Label description, and the span of the label.
    labels: Vec<(Vec<SubMsg<'i>>, &'i str)>,
    /// Notes and their spans.
    notes: Vec<(Vec<SubMsg<'i>>, Option<&'i str>)>,
    /// Helps and their spans.
    /// Helps are additional information that can help the user understand the error.
    helps: Vec<(Vec<SubMsg<'i>>, Option<&'i str>)>,
    lint: &'static Lint,
}

declare_arena!(
    [
        [] _phantom: &'tcx (),
    ]
);

static ARENA: LazyLock<Arena<'static>> = LazyLock::new(Arena::default);

impl<'i> DynamicErrorBuilder<'i> {
    //FIXME: this function has a lot of `unwrap` calls, which can panic if the input is malformed.
    /// Create a [`DynamicErrorBuilder`] from a [`pairs::diagBlockItem`].
    pub(super) fn from_item(item: &'i pairs::diagBlockItem<'i>, meta_vars: &NonLocalMetaSymTab) -> Self {
        let (_, _, _, pairs, _, _) = item.get_matched();
        let mut primary = (None, None);
        let mut labels = Vec::new();
        let mut notes = Vec::new();
        let mut helps = Vec::new();
        let mut level = Level::Deny;
        let mut name = None;

        for pair in pairs.iter_matched() {
            let (key, span, _, message, _) = pair.get_matched();

            let message = message.get_matched().1;

            let key = key.span.as_str();

            let span_name = span.as_ref().map(|span| span.get_matched().1.span.as_str());

            match key {
                "primary" => {
                    primary.0 = Some(SubMsg::parse(message, meta_vars));
                    if let Some(span_name) = span_name {
                        primary.1 = Some(span_name);
                    }
                },
                "label" => {
                    labels.push((SubMsg::parse(message, meta_vars), span_name.unwrap()));
                },
                "note" => {
                    notes.push((SubMsg::parse(message, meta_vars), span_name));
                },
                "help" => {
                    helps.push((SubMsg::parse(message, meta_vars), span_name));
                },
                "name" => {
                    name = Some(message);
                },
                "level" => {
                    let message = message.span.as_str();
                    level = match message {
                        "allow" => Level::Allow,
                        "warning" => Level::Warn,
                        "deny" => Level::Deny,
                        "forbid" => Level::Forbid,
                        _ => unimplemented!("Unrecognized level: {message}",),
                    };
                },
                _ => unimplemented!("Unrecognized key: {key:?}"),
            }
        }
        let primary = (primary.0.unwrap(), primary.1.unwrap());
        let name = name.unwrap().span.as_str();
        let builder = DynamicErrorBuilder {
            primary,
            labels,
            notes,
            helps,
            lint: ARENA.alloc(Lint {
                name: ARENA.alloc_str(&format!("rpl::{name}")),
                default_level: level,
                ..Lint::default_fields_for_macro()
            }),
        };
        builder
    }
    pub(crate) fn build<'tcx>(
        &self,
        label_map: &LabelMap,
        body: &Body<'tcx>,
        matched: &impl Matched<'tcx>,
    ) -> DynamicError {
        fn format<'tcx>(message: &Vec<SubMsg>, matched: &impl Matched<'tcx>) -> String {
            let mut s = String::new();
            for msg in message {
                match msg {
                    SubMsg::Str(smsg) => s.push_str(smsg),
                    SubMsg::Ty(idx) => {
                        let ty = matched.type_meta_var(*idx);
                        s.push_str(&ty.to_string());
                    },
                    SubMsg::Const(idx) => {
                        let const_ = matched.const_meta_var(*idx);
                        s.push_str(&const_.to_string());
                    },
                }
            }
            s
        }
        let primary = (
            format(&self.primary.0, matched),
            matched.span(label_map, body, self.primary.1),
        );
        let labels = self
            .labels
            .iter()
            .map(|(label, span)| (format(label, matched), matched.span(label_map, body, span)))
            .collect();
        let notes = self
            .notes
            .iter()
            .map(|(note, span)| {
                (
                    format(note, matched),
                    span.map(|span| matched.span(label_map, body, span)),
                )
            })
            .collect();
        let helps = self
            .helps
            .iter()
            .map(|(help, span)| {
                (
                    format(help, matched),
                    span.map(|span| matched.span(label_map, body, span)),
                )
            })
            .collect();
        let lint = self.lint;
        DynamicError {
            primary,
            labels,
            notes,
            helps,
            lint,
        }
    }
}
