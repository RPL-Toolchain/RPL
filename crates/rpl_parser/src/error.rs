use crate::SpanWrapper;
use crate::parser::Rule;
use crate::position::PositionWrapper;
use colored::Colorize;
use derive_more::derive::Debug;
use pest_typed::tracker::{SpecialError, Tracker};
use pest_typed::{Input, Position};
use rustc_errors::{Diag, DiagCtxtHandle, Diagnostic, Level};
use rustc_span::ErrorGuaranteed;
use std::collections::BTreeMap;
use std::convert::identity;
use std::fmt::{Display, Write};
use std::path::Path;

fn format_rule(rule: Option<Rule>, f: &mut impl Write) -> std::fmt::Result {
    match rule {
        None => write!(f, "Root Rule")?,
        Some(rule) => write!(f, "{:?}", rule)?,
    }
    Ok(())
}

fn format_tracker<'i>(
    path: &'i Path,
    pos: Position<'i>,
    rules: &BTreeMap<Option<Rule>, Tracked<Rule>>,
    replacer: impl FnOnce(pest_typed::Position<'i>) -> pest_typed::Position<'i>,
    f: &mut impl Write,
) -> Result<(), std::fmt::Error> {
    let pos = replacer(pos);
    write!(f, "{}", PositionWrapper::new(pos, path))?;
    let log10 = {
        let mut n = pos.line_col().0;
        let mut i = 1;
        while n >= 10 {
            n /= 10;
            i += 1;
        }
        i
    };

    write!(f, "\n{} {}", " ".repeat(log10), "Possible RPL grammar rules: ".blue())?;
    write!(f, "[")?;
    let mut iter = rules.keys().cloned();
    if let Some(rule) = iter.next() {
        format_rule(rule, f)?;
    }
    for rule in iter {
        write!(f, ",")?;
        format_rule(rule, f)?;
    }
    write!(f, "]")?;

    Ok(())
}

type Tracked<R> = (Vec<R>, Vec<R>, Vec<SpecialError>);

/// Errors from parser.
#[derive(Debug)]
#[debug("ParseError({path:?})")]
pub struct ParseError<'i> {
    path: &'i Path,
    position: Position<'i>,
    attempts: BTreeMap<Option<Rule>, Tracked<Rule>>,
}

impl<'i> ParseError<'i> {
    /// Create a [ParseError].'
    pub fn new(tracker: Tracker<'i, Rule>, path: &'i Path) -> Self {
        let (pos, rules) = tracker.finish();
        Self {
            path,
            position: pos,
            attempts: rules,
        }
    }
}

impl<'i> Display for ParseError<'i> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        format_tracker(self.path, self.position, &self.attempts, identity, f)
    }
}
impl std::error::Error for ParseError<'_> {}

struct Attempts<'e>(&'e BTreeMap<Option<Rule>, Tracked<Rule>>);

impl Display for Attempts<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "[")?;
        let mut iter = self.0.keys();
        if let Some(rule) = iter.next() {
            format_rule(*rule, f)?;
        }
        for rule in iter {
            write!(f, ", ")?;
            format_rule(*rule, f)?;
        }
        write!(f, "]")?;
        Ok(())
    }
}

impl Diagnostic<'_, ErrorGuaranteed> for &ParseError<'_> {
    fn into_diag(self, dcx: DiagCtxtHandle<'_>, _: Level) -> Diag<'_, ErrorGuaranteed> {
        let span = self.position.span(
            &Position::new(
                self.position.input(),
                self.position.pos()
                    + self.position.input()[self.position.pos()..]
                        .chars()
                        .next()
                        .map(|c| c.len_utf8())
                        .unwrap_or(0),
            )
            .unwrap(),
        );
        dcx.struct_err("Parse error")
            .with_help(format!("Possible RPL grammar rules: {}", Attempts(&self.attempts)))
            .with_help(SpanWrapper::new(span, self.path).to_string())
    }
}

// //FIXME: the source file it uses is incorrect
// pub fn span_cvt(span: SpanWrapper<'_>) -> rustc_span::Span {
//     let expn_id = LocalExpnId::fresh_empty();
//     let expn_data = ExpnData::default(
//         ExpnKind::AstPass(AstPass::StdImports),
//         Span::new(BytePos(0), BytePos(0), SyntaxContext::root(), None),
//         LATEST_STABLE_EDITION,
//         None,
//         None,
//     );
//     expn_id.set_expn_data(expn_data, ctx);
//     let expn_id = expn_id.to_expn_id();
//     let ctx = SyntaxContext::root().apply_mark(expn_id, Transparency::Opaque);
//     rustc_span::Span::new(
//         rustc_span::BytePos(span.inner().start().try_into().unwrap()),
//         rustc_span::BytePos(span.inner().end().try_into().unwrap()),
//         ctx,
//         None,
//     )
// }
