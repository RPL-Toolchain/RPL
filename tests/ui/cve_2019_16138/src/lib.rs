//@compile-flags: -Z inline-mir=false
extern crate scoped_threadpool;

pub use hdr::decoder::HDRDecoder;

mod color {
    /// An enumeration over supported color types and their bit depths
    #[derive(Copy, PartialEq, Eq, Debug, Clone, Hash)]
    pub enum ColorType {
        /// Pixel is grayscale
        Gray(u8),

        /// Pixel contains R, G and B channels
        RGB(u8),

        /// Pixel is an index into a color palette
        Palette(u8),

        /// Pixel is grayscale with an alpha channel
        GrayA(u8),

        /// Pixel is RGB with an alpha channel
        RGBA(u8),

        /// Pixel contains B, G and R channels
        BGR(u8),

        /// Pixel is BGR with an alpha channel
        BGRA(u8),
    }
}

mod image {
    use std::error::Error;
    use std::fmt;
    use std::io;

    use crate::color::ColorType;

    /// An enumeration of Image errors
    #[derive(Debug)]
    pub enum ImageError {
        /// The Image is not formatted properly
        FormatError(String),

        /// The Image's dimensions are either too small or too large
        DimensionError,

        /// The Decoder does not support this image format
        UnsupportedError(String),

        /// The Decoder does not support this color type
        UnsupportedColor(ColorType),

        /// Not enough data was provided to the Decoder
        /// to decode the image
        NotEnoughData,

        /// An I/O Error occurred while decoding the image
        IoError(io::Error),

        /// The end of the image has been reached
        ImageEnd,

        /// There is not enough memory to complete the given operation
        InsufficientMemory,
    }

    impl fmt::Display for ImageError {
        fn fmt(&self, fmt: &mut fmt::Formatter) -> Result<(), fmt::Error> {
            match *self {
                ImageError::FormatError(ref e) => write!(fmt, "Format error: {}", e),
                ImageError::DimensionError => write!(
                    fmt,
                    "The Image's dimensions are either too \
                     small or too large"
                ),
                ImageError::UnsupportedError(ref f) => write!(
                    fmt,
                    "The Decoder does not support the \
                     image format `{}`",
                    f
                ),
                ImageError::UnsupportedColor(ref c) => write!(
                    fmt,
                    "The decoder does not support \
                     the color type `{:?}`",
                    c
                ),
                ImageError::NotEnoughData => write!(
                    fmt,
                    "Not enough data was provided to the \
                     Decoder to decode the image"
                ),
                ImageError::IoError(ref e) => e.fmt(fmt),
                ImageError::ImageEnd => write!(fmt, "The end of the image has been reached"),
                ImageError::InsufficientMemory => write!(fmt, "Insufficient memory"),
            }
        }
    }

    impl Error for ImageError {
        fn description(&self) -> &str {
            match *self {
                ImageError::FormatError(..) => "Format error",
                ImageError::DimensionError => "Dimension error",
                ImageError::UnsupportedError(..) => "Unsupported error",
                ImageError::UnsupportedColor(..) => "Unsupported color",
                ImageError::NotEnoughData => "Not enough data",
                ImageError::IoError(..) => "IO error",
                ImageError::ImageEnd => "Image end",
                ImageError::InsufficientMemory => "Insufficient memory",
            }
        }

        fn cause(&self) -> Option<&dyn Error> {
            match *self {
                ImageError::IoError(ref e) => Some(e),
                _ => None,
            }
        }
    }

    impl From<io::Error> for ImageError {
        fn from(err: io::Error) -> ImageError {
            ImageError::IoError(err)
        }
    }

    /// Result of an image decoding/encoding process
    pub type ImageResult<T> = Result<T, ImageError>;
}

mod hdr {
    pub(super) mod decoder {
        use scoped_threadpool::Pool;
        #[cfg(test)]
        use std::borrow::Cow;
        use std::io::{self, BufRead};
        use std::iter::Iterator;

        use crate::image::{ImageError, ImageResult};

        /// An Radiance HDR decoder
        #[derive(Debug)]
        pub struct HDRDecoder<R> {
            r: R,
            width: u32,
            height: u32,
            meta: HDRMetadata,
        }

        /// Refer to [wikipedia](https://en.wikipedia.org/wiki/RGBE_image_format)
        #[repr(C)]
        #[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
        pub struct RGBE8Pixel {
            /// Color components
            pub c: [u8; 3],
            /// Exponent
            pub e: u8,
        }

        impl<R: BufRead> HDRDecoder<R> {
            /// Returns file metadata. Refer to ```HDRMetadata``` for details.
            pub fn metadata(&self) -> HDRMetadata {
                self.meta.clone()
            }

            /// Consumes decoder and returns a vector of RGBE8 pixels
            pub fn read_image_native(mut self) -> ImageResult<Vec<RGBE8Pixel>> {
                // Don't read anything if image is empty
                if self.width == 0 || self.height == 0 {
                    return Ok(vec![]);
                }
                // expression self.width > 0 && self.height > 0 is true from now to the end of this method
                let pixel_count = self.width as usize * self.height as usize;
                let mut ret = vec![Default::default(); pixel_count];
                for chunk in ret.chunks_mut(self.width as usize) {
                    read_scanline(&mut self.r, chunk)?;
                }
                Ok(ret)
            }

            /// Consumes decoder and returns a vector of transformed pixels
            pub fn read_image_transform<T: Send, F: Send + Sync + Fn(RGBE8Pixel) -> T>(
                mut self,
                f: F,
            ) -> ImageResult<Vec<T>> {
                // Don't read anything if image is empty
                if self.width == 0 || self.height == 0 {
                    return Ok(vec![]);
                }
                // expression self.width > 0 && self.height > 0 is true from now to the end of this method
                // scanline buffer
                let uszwidth = self.width as usize;

                let pixel_count = self.width as usize * self.height as usize;
                let mut ret = Vec::with_capacity(pixel_count);
                unsafe {
                    // RGBE8Pixel doesn't implement Drop, so it's Ok to drop half-initialized ret
                    ret.set_len(pixel_count);
                } // ret contains uninitialized data, so now it's my responsibility to return fully initialized ret

                {
                    let chunks_iter = ret.chunks_mut(uszwidth);
                    let mut pool = Pool::new(8); //

                    (pool.scoped(|scope| {
                        for chunk in chunks_iter {
                            let mut buf = Vec::<RGBE8Pixel>::with_capacity(uszwidth);
                            unsafe {
                                buf.set_len(uszwidth);
                            }
                            (read_scanline(&mut self.r, &mut buf[..]))?;
                            let f = &f;
                            scope.execute(move || {
                                for (dst, &pix) in chunk.iter_mut().zip(buf.iter()) {
                                    *dst = f(pix);
                                }
                            });
                        }
                        Ok(())
                    }) as Result<(), ImageError>)?;
                }

                Ok(ret)
            }
        }

        impl<R: BufRead> IntoIterator for HDRDecoder<R> {
            type Item = ImageResult<RGBE8Pixel>;
            type IntoIter = HDRImageDecoderIterator<R>;

            fn into_iter(self) -> Self::IntoIter {
                HDRImageDecoderIterator {
                    r: self.r,
                    scanline_cnt: self.height as usize,
                    buf: vec![Default::default(); self.width as usize],
                    col: 0,
                    scanline: 0,
                    trouble: true, // make first call to `next()` read scanline
                    error_encountered: false,
                }
            }
        }

        /// Scanline buffered pixel by pixel iterator
        pub struct HDRImageDecoderIterator<R: BufRead> {
            r: R,
            scanline_cnt: usize,
            buf: Vec<RGBE8Pixel>, // scanline buffer
            col: usize,           // current position in scanline
            scanline: usize,      // current scanline
            trouble: bool,        // optimization, true indicates that we need to check something
            error_encountered: bool,
        }

        impl<R: BufRead> HDRImageDecoderIterator<R> {
            // Advances counter to the next pixel
            #[inline]
            fn advance(&mut self) {
                self.col += 1;
                if self.col == self.buf.len() {
                    self.col = 0;
                    self.scanline += 1;
                    self.trouble = true;
                }
            }
        }

        impl<R: BufRead> Iterator for HDRImageDecoderIterator<R> {
            type Item = ImageResult<RGBE8Pixel>;

            fn next(&mut self) -> Option<Self::Item> {
                if !self.trouble {
                    let ret = self.buf[self.col];
                    self.advance();
                    Some(Ok(ret))
                } else {
                    // some condition is pending
                    if self.buf.is_empty() || self.scanline == self.scanline_cnt {
                        // No more pixels
                        return None;
                    } // no else
                    if self.error_encountered {
                        self.advance();
                        // Error was encountered. Keep producing errors.
                        // ImageError can't implement Clone, so just dump some error
                        return Some(Err(ImageError::ImageEnd));
                    } // no else
                    if self.col == 0 {
                        // fill scanline buffer
                        match read_scanline(&mut self.r, &mut self.buf[..]) {
                            Ok(_) => {
                                // no action required
                            }
                            Err(err) => {
                                self.advance();
                                self.error_encountered = true;
                                self.trouble = true;
                                return Some(Err(err));
                            }
                        }
                    } // no else
                    self.trouble = false;
                    let ret = self.buf[0];
                    self.advance();
                    Some(Ok(ret))
                }
            }

            fn size_hint(&self) -> (usize, Option<usize>) {
                let total_cnt = self.buf.len() * self.scanline_cnt;
                let cur_cnt = self.buf.len() * self.scanline + self.col;
                let remaining = total_cnt - cur_cnt;
                (remaining, Some(remaining))
            }
        }

        impl<R: BufRead> ExactSizeIterator for HDRImageDecoderIterator<R> {}

        // Precondition: buf.len() > 0
        fn read_scanline<R: BufRead>(r: &mut R, buf: &mut [RGBE8Pixel]) -> ImageResult<()> {
            assert!(!buf.is_empty());
            let width = buf.len();
            // first 4 bytes in scanline allow to determine compression method
            let fb = read_rgbe(r)?;
            if fb.c[0] == 2 && fb.c[1] == 2 && fb.c[2] < 128 {
                // denormalized pixel value (2,2,<128,_) indicates new per component RLE method
                // decode_component guarantees that offset is within 0 .. width
                // therefore we can skip bounds checking here, but we will not
                decode_component(r, width, |offset, value| buf[offset].c[0] = value)?;
                decode_component(r, width, |offset, value| buf[offset].c[1] = value)?;
                decode_component(r, width, |offset, value| buf[offset].c[2] = value)?;
                decode_component(r, width, |offset, value| buf[offset].e = value)?;
            } else {
                // old RLE method (it was considered old around 1991, should it be here?)
                decode_old_rle(r, fb, buf)?;
            }
            Ok(())
        }

        #[inline(always)]
        fn read_byte<R: BufRead>(r: &mut R) -> io::Result<u8> {
            let mut buf = [0u8];
            r.read_exact(&mut buf[..])?;
            Ok(buf[0])
        }

        // Guarantees that first parameter of set_component will be within pos .. pos+width
        #[inline]
        fn decode_component<R: BufRead, S: FnMut(usize, u8)>(
            r: &mut R,
            width: usize,
            mut set_component: S,
        ) -> ImageResult<()> {
            let mut buf = [0; 128];
            let mut pos = 0;
            while pos < width {
                // increment position by a number of decompressed values
                pos += {
                    let rl = read_byte(r)?;
                    if rl <= 128 {
                        // sanity check
                        if pos + rl as usize > width {
                            return Err(ImageError::FormatError(
                                "Wrong length of decoded scanline".into(),
                            ));
                        }
                        // read values
                        r.read_exact(&mut buf[0..rl as usize])?;
                        for (offset, &value) in buf[0..rl as usize].iter().enumerate() {
                            set_component(pos + offset, value);
                        }
                        rl as usize
                    } else {
                        // ?run
                        let rl = rl - 128;
                        // sanity check
                        if pos + rl as usize > width {
                            return Err(ImageError::FormatError(
                                "Wrong length of decoded scanline".into(),
                            ));
                        }
                        // fill with same value
                        let value = read_byte(r)?;
                        for offset in 0..rl as usize {
                            set_component(pos + offset, value);
                        }
                        rl as usize
                    }
                };
            }
            if pos != width {
                return Err(ImageError::FormatError(
                    "Wrong length of decoded scanline".into(),
                ));
            }
            Ok(())
        }

        // Decodes scanline, places it into buf
        // Precondition: buf.len() > 0
        // fb - first 4 bytes of scanline
        fn decode_old_rle<R: BufRead>(
            r: &mut R,
            fb: RGBE8Pixel,
            buf: &mut [RGBE8Pixel],
        ) -> ImageResult<()> {
            assert!(!buf.is_empty());
            let width = buf.len();
            // convenience function.
            // returns run length if pixel is a run length marker
            #[inline]
            fn rl_marker(pix: RGBE8Pixel) -> Option<usize> {
                if pix.c == [1, 1, 1] {
                    Some(pix.e as usize)
                } else {
                    None
                }
            }
            // first pixel in scanline should not be run length marker
            // it is error if it is
            if rl_marker(fb).is_some() {
                return Err(ImageError::FormatError(
                    "First pixel of a scanline shouldn't be run length marker".into(),
                ));
            }
            buf[0] = fb; // set first pixel of scanline

            let mut x_off = 1; // current offset from beginning of a scanline
            let mut rl_mult = 1; // current run length multiplier
            let mut prev_pixel = fb;
            while x_off < width {
                let pix = read_rgbe(r)?;
                // it's harder to forget to increase x_off if I write this this way.
                x_off += {
                    if let Some(rl) = rl_marker(pix) {
                        // rl_mult takes care of consecutive RL markers
                        let rl = rl * rl_mult;
                        rl_mult *= 256;
                        if x_off + rl <= width {
                            // do run
                            for b in &mut buf[x_off..x_off + rl] {
                                *b = prev_pixel;
                            }
                        } else {
                            return Err(ImageError::FormatError(
                                "Wrong length of decoded scanline".into(),
                            ));
                        };
                        rl // value to increase x_off by
                    } else {
                        rl_mult = 1; // chain of consecutive RL markers is broken
                        prev_pixel = pix;
                        buf[x_off] = pix;
                        1 // value to increase x_off by
                    }
                };
            }
            if x_off != width {
                return Err(ImageError::FormatError(
                    "Wrong length of decoded scanline".into(),
                ));
            }
            Ok(())
        }

        fn read_rgbe<R: BufRead>(r: &mut R) -> io::Result<RGBE8Pixel> {
            let mut buf = [0u8; 4];
            r.read_exact(&mut buf[..])?;
            Ok(RGBE8Pixel {
                c: [buf[0], buf[1], buf[2]],
                e: buf[3],
            })
        }

        /// Metadata for Radiance HDR image
        #[derive(Debug, Clone)]
        pub struct HDRMetadata {
            /// Width of decoded image. It could be either scanline length,
            /// or scanline count, depending on image orientation.
            pub width: u32,
            /// Height of decoded image. It depends on orientation too.
            pub height: u32,
            /// Orientation matrix. For standard orientation it is ((1,0),(0,1)) - left to right, top to bottom.
            /// First pair tells how resulting pixel coordinates change along a scanline.
            /// Second pair tells how they change from one scanline to the next.
            pub orientation: ((i8, i8), (i8, i8)),
            /// Divide color values by exposure to get to get physical radiance in
            /// watts/steradian/m<sup>2</sup>
            ///
            /// Image may not contain physical data, even if this field is set.
            pub exposure: Option<f32>,
            /// Divide color values by corresponding tuple member (r, g, b) to get to get physical radiance
            /// in watts/steradian/m<sup>2</sup>
            ///
            /// Image may not contain physical data, even if this field is set.
            pub color_correction: Option<(f32, f32, f32)>,
            /// Pixel height divided by pixel width
            pub pixel_aspect_ratio: Option<f32>,
            /// All lines contained in image header are put here. Ordering of lines is preserved.
            /// Lines in the form "key=value" are represented as ("key", "value").
            /// All other lines are ("", "line")
            pub custom_attributes: Vec<(String, String)>,
        }
    }
}
