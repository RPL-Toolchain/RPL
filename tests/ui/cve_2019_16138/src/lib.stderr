error: Use `Vec::set_len` to extend the length of a `Vec`, potentially including uninitialized elements
  --> tests/ui/cve_2019_16138/src/lib.rs:200:25
   |
LL |                 let mut ret = Vec::with_capacity(pixel_count);
   |                               ------------------------------- `Vec` created here
...
LL |                     ret.set_len(pixel_count);
   |                         ^^^^^^^^^^^^^^^^^^^^ `Vec::set_len` called here
   |
   = note: make sure all elements are initialized before using them
   = note: `#[deny(rpl::set_len_to_extend)]` on by default

error: it violates the precondition of `Vec::set_len` to extend a `Vec`'s length without initializing its content in advance
  --> tests/ui/cve_2019_16138/src/lib.rs:200:25
   |
LL |                 let mut ret = Vec::with_capacity(pixel_count);
   |                               ------------------------------- `Vec` created here
...
LL |                     ret.set_len(pixel_count);
   |                         ^^^^^^^^^^^^^^^^^^^^ `Vec::set_len` called here
   |
   = help: before calling `set_len` to extend its length, make sure all elements are initialized, using such as `spare_capacity_mut` or `as_mut_ptr`
   = note: `#[deny(rpl::set_len_uninitialized)]` on by default

error: it violates the precondition of `Vec::set_len` to extend a `Vec`'s length without initializing its content in advance
  --> tests/ui/cve_2019_16138/src/lib.rs:215:37
   |
LL | ...                   let mut buf = Vec::<RGBE8Pixel>::with_capacity(uszwidth);
   |                                     ------------------------------------------ `Vec` created here
LL | ...                   unsafe {
LL | ...                       buf.set_len(uszwidth);
   |                               ^^^^^^^^^^^^^^^^^ `Vec::set_len` called here
   |
   = help: before calling `set_len` to extend its length, make sure all elements are initialized, using such as `spare_capacity_mut` or `as_mut_ptr`

error: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
  --> tests/ui/cve_2019_16138/src/lib.rs:206:39
   |
LL |                 let mut ret = Vec::with_capacity(pixel_count);
   |                               ------------------------------- `std::vec::Vec` created but not initialized
...
LL |                     let chunks_iter = ret.chunks_mut(uszwidth);
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^
   |                                       |
   |                                       slice created here
   |                                       slice created with this length
   |                                       slice created with this pointer
   |
   = help: See https://doc.rust-lang.org/std/slice/fn.std::slice::from_raw_parts_mut.html
   = note: `#[deny(rpl::slice_from_raw_parts_uninitialized)]` on by default

error: it violates the precondition of `std::slice::from_raw_parts_mut` to create a slice from uninitialized data
  --> tests/ui/cve_2019_16138/src/lib.rs:218:65
   |
LL | ...                   let mut buf = Vec::<RGBE8Pixel>::with_capacity(uszwidth);
   |                                     ------------------------------------------ `std::vec::Vec` created but not initialized
...
LL | ...                   (read_scanline(&mut self.r, &mut buf[..]))?;
   |                                                           ^^^^
   |                                                           |
   |                                                           slice created here
   |                                                           slice created with this length
   |                                                           slice created with this pointer
   |
   = help: See https://doc.rust-lang.org/std/slice/fn.std::slice::from_raw_parts_mut.html

error: it usually isn’t necessary to apply #[inline] to private functions
  --> tests/ui/cve_2019_16138/src/lib.rs:267:13
   |
LL | /             fn advance(&mut self) {
LL | |
LL | |                 self.col += 1;
LL | |                 if self.col == self.buf.len() {
...  |
LL | |             }
   | |_____________^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html
   = note: within a crate, the compiler generally makes good inline decisions
   = note: `#[deny(rpl::private_and_inline)]` on by default

error: it usually isn’t necessary to apply #[inline] to private functions
  --> tests/ui/cve_2019_16138/src/lib.rs:351:9
   |
LL | /         fn read_byte<R: BufRead>(r: &mut R) -> io::Result<u8> {
LL | |
LL | |             let mut buf = [0u8];
LL | |             r.read_exact(&mut buf[..])?;
LL | |             Ok(buf[0])
LL | |         }
   | |_________^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html
   = note: within a crate, the compiler generally makes good inline decisions

error: it usually isn’t necessary to apply #[inline] to private functions
  --> tests/ui/cve_2019_16138/src/lib.rs:360:9
   |
LL | /         fn decode_component<R: BufRead, S: FnMut(usize, u8)>(
LL | |
LL | |             r: &mut R,
LL | |             width: usize,
...  |
LL | |             Ok(())
LL | |         }
   | |_________^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html
   = note: within a crate, the compiler generally makes good inline decisions

error: it usually isn’t necessary to apply #[inline] to private functions
  --> tests/ui/cve_2019_16138/src/lib.rs:424:13
   |
LL | /             fn rl_marker(pix: RGBE8Pixel) -> Option<usize> {
LL | |
LL | |
LL | |                 if pix.c == [1, 1, 1] {
...  |
LL | |             }
   | |_____________^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html
   = note: within a crate, the compiler generally makes good inline decisions

error: it usually isn’t necessary to apply #[inline] to private functions
  --> tests/ui/cve_2019_16138/src/lib.rs:424:13
   |
LL | /             fn rl_marker(pix: RGBE8Pixel) -> Option<usize> {
LL | |
LL | |
LL | |                 if pix.c == [1, 1, 1] {
...  |
LL | |             }
   | |_____________^ `#[inline]` applied here
   |
   = help: See https://matklad.github.io/2021/07/09/inline-in-rust.html
   = note: within a crate, the compiler generally makes good inline decisions
   = note: duplicate diagnostic emitted due to `-Z deduplicate-diagnostics=no`

error: aborting due to 10 previous errors

